.travis.yml

```diff
@@ -39,10 +41,11 @@ install:
   - npm install -g jsonlint
 
 script:
-  - find . -type f -name "*.json" -print0 | xargs -0 -n1 jsonlint -q
-  - find . -type f -name "*.php" -print0 | xargs -0 -n1 php -l
-  - composer validate
+  - if [[ "$SNIFF" == "1" ]];then find . -type f -name "*.json" -print0 | xargs -0 -n1 jsonlint -q; fi
+  - if [[ "$SNIFF" == "1" ]];then find . -type f -name "*.php" -print0 | xargs -0 -n1 php -l; fi
+  - if [[ "$SNIFF" == "1" ]];then composer validate; fi
   - travis_wait composer install --no-interaction --no-progress --no-scripts --no-suggest --optimize-autoloader --prefer-dist --verbose
-  - if [[ "$SNIFF" == "1" ]]; then composer install-codestandards; fi
-  - if [[ "$SNIFF" == "1" ]]; then ./vendor/bin/phpcs; fi
-  - ./vendor/bin/security-checker -n security:check --end-point=http://security.sensiolabs.org/check_lock
+  - composer install-codestandards
+  - ./vendor/bin/phpcs -i
+  - ./vendor/bin/phpcs --standard=PHPCompatibility --runtime-set testVersion "${TRAVIS_PHP_VERSION:0:3}" --exclude=PHPCompatibility.Upgrade.LowPHP ./src/
```

Comment:
The sniffs will return the same results cross-version (and tested for this in PHPCompatibility itself), so you're now effectively unit testing PHPCompatibility, which you don't need to do.

Reply:
@jrfnl i don't follow, so can you elaborate please?

what this is doing afaik is running it phpcs in 'run in php version X' which should give different warnings/errors depending on when a featured was added or removed in php

Reply:
@mjrider This command (as-is) is running the PHPCompatibility sniffs against each PHP version and with a `testVersion` which is the same as the PHP version the command is run on.

PHPCompatibility, however, supports `testVersion` ranges, so, running it once on a high PHP version with a `testVersion` range will get you the same or even better results (some issue will only display when you want to support a range of PHP versions).

`./vendor/bin/phpcs --standard=PHPCompatibility --runtime-set testVersion 5.3- --exclude=PHPCompatibility.Upgrade.LowPHP ./src/`

Ref: https://github.com/PHPCompatibility/PHPCompatibility#sniffing-your-code-for-compatibility-with-specific-php-versions

Better yet, run the normal CS check (as mentioned below) and it includes the PHPCompatibility check with the correct `testVersion`.

See: https://github.com/Dealerdirect/phpcodesniffer-composer-installer/blob/master/phpcs.xml.dist#L15-L16

Reply:
i see what you are doing,
by running it per every version, you get a specific build to fail. so when there is a change breaking php 5.6 but not 5.5 and 7.0 only the build for php 5.6 breaks, so without even reading the build output, you already know which php version is breaking.

with this pipeline, it is checking php 7.4 with allow failure,
so effectively we are running it for 5.3-7.3 and 7.4 but we do not care about it complaining.

imho it's a matter of taste if you take one way or the other. Personally i would run it for every version separately

Reply:
> imho it's a matter of taste if you take one way or the other

It's not a matter of taste. It's a matter of understanding how this works. I clearly did not explain this well enough.

You can have separate runs with a different fixed `testVersion` setting, but...
1. The results will be **_less_ accurate** compared to running the sniffs once with a `testVersion` range.
    As I said above, some sniffs will only trigger when a combination of PHP versions needs to be supported.
    For example: PHP changed the default charset for some functions in PHP 5.4. If you only run with `5.3` or `5.4` as `testVersion` you will not receive an error for that, as when only that version needs to be supported, it doesn't cause a problem.
    However, if you run with a `testVersion` of `5.3-5.4`, you will receive an error.
2. While generally speaking PHPCompatibility will give the same results (given the same `testVersion`) across all PHP versions on which PHPCompatibility is supported, there are some tokenizer quirks which can't be worked-around.
    So for the _best_ results, it is strongly recommended to run on a high PHP version.